runtime/src/main/scala/grpc/akkastreams/GrpcSourceStage.scala

```diff
@@ -0,0 +1,43 @@
+package grpc.akkastreams
+
+import akka.stream.{Attributes, Outlet, SourceShape}
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, OutHandler}
+import io.grpc.stub.{CallStreamObserver, StreamObserver}
+
+import scala.concurrent.{Future, Promise}
+
+class GrpcSourceStage[I, O](requestStream: CallStreamObserver[O])
+  extends GraphStageWithMaterializedValue[SourceShape[I], Future[StreamObserver[I]]] {
+  val out = Outlet[I]("grpc.out")
+  override val shape: SourceShape[I] = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(
+    inheritedAttributes: Attributes
+  ): (GraphStageLogic, Future[StreamObserver[I]]) = {
+    val promise: Promise[StreamObserver[I]] = Promise()
+
+    val logic = new GraphStageLogic(shape) with OutHandler {
+      val inObs = new StreamObserver[I] {
+        override def onError(t: Throwable) =
+          getAsyncCallback((t: Throwable) => fail(out, t)).invoke(t)
+
+        override def onCompleted() =
+          getAsyncCallback((_: Unit) => complete(out)).invoke(())
+
+        override def onNext(value: I) =
+          getAsyncCallback((value: I) => emit(out, value)).invoke(value)
```

Comment:
`push` should be fine instead of `emit`; emit is a more advanced tool that you're not really making use of there (it can model state transitions)